packages/common/src/queue.test.ts

```diff
@@ -0,0 +1,349 @@
+
+test("start", async () => {
+  const queue = createQueue({
+    concurrency: 1,
+    frequency: 1,
+    worker: () => Promise.resolve(),
+  });
+
+  const promise = queue.add();
+
+  expect(queue.size()).toBe(1);
+
+  queue.start();
+
+  expect(queue.isStarted()).toBe(true);
+
+  await promise;
+
+  expect(queue.size()).toBe(0);
+});
+
+test("pause", () => {
+  const queue = createQueue({
+    concurrency: 1,
+    frequency: 1,
+    worker: () => Promise.resolve(),
+  });
+
+  queue.start();
+  queue.pause();
+
+  queue.add();
+
+  expect(queue.size()).toBe(1);
+});
+
+test("onIdle short loop", async () => {
+  const queue = createQueue({
+    concurrency: 1,
+    frequency: 1,
+    worker: () => Promise.resolve(),
+  });
+
+  await queue.onIdle();
+});
+
+test("onIdle", async () => {
+  const queue = createQueue({
+    concurrency: 1,
+    worker: () => Promise.resolve(),
+  });
+
+  queue.add();
+
+  const promise = queue.onIdle();
+
+  queue.start();
+
+  await promise;
+});
+
+test("onIdle twice", async () => {
+  const queue = createQueue({
+    concurrency: 1,
+    frequency: 1,
+    worker: () => Promise.resolve(),
+  });
+
+  queue.add();
+
+  queue.onIdle();
+
+  queue.start();
+
+  queue.pause();
+
+  queue.add();
+
+  const promise = queue.onIdle();
+
+  queue.start();
+
+  await promise;
+});
+
+test("onEmpty short loop", async () => {
+  const queue = createQueue({
+    concurrency: 1,
+    frequency: 1,
+    worker: () => Promise.resolve(),
+  });
+
+  await queue.onEmpty();
+});
+
+test("onEmpty", async () => {
+  const queue = createQueue({
+    concurrency: 1,
+    frequency: 1,
+    worker: () => Promise.resolve(),
+  });
+
+  queue.add();
+
+  const promise = queue.onEmpty();
+
+  queue.start();
+
+  await promise;
+});
+
+test("onEmpty twice", async () => {
+  const queue = createQueue({
+    concurrency: 1,
+    frequency: 1,
+    worker: () => Promise.resolve(),
+  });
+
+  queue.add();
+
+  queue.onEmpty();
+
+  queue.start();
+
+  queue.pause();
+
+  queue.add();
+
+  const promise = queue.onEmpty();
+
+  queue.start();
+
+  await promise;
+});
+
+test("concurrency", () => {
+  const func = vi.fn(() => Promise.resolve());
+
+  const queue = createQueue({
+    concurrency: 2,
+    frequency: 5,
+    worker: func,
+  });
+
+  queue.add();
+  queue.add();
+  queue.add();
+  queue.add();
+
+  queue.start();
+  queue.pause();
+
+  expect(queue.size()).toBe(2);
+  expect(func).toHaveBeenCalledTimes(2);
+});
+
+test("frequency", async () => {
+  const func = vi.fn(() => Promise.resolve());
+
+  const queue = createQueue({
+    frequency: 2,
+    concurrency: 5,
+    worker: func,
+  });
+
+  queue.add();
+  queue.add();
+  queue.add();
+  queue.add();
+
+  queue.start();
+
+  expect(queue.size()).toBe(2);
+  expect(func).toHaveBeenCalledTimes(2);
+
+  await new Promise((resolve) => setTimeout(resolve, 1_010));
+
+  expect(queue.size()).toBe(0);
+  expect(func).toHaveBeenCalledTimes(4);
+});
+
+test("event loop", async () => {
```

Comment:
Not sure what this is testing - could you write a bit more in the test description?